include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+        // every group of 64 invocations has been coallesced into 1 result value. This means that the results of 
+        // the first SubgroupSz^2 invocations will be processed by the first subgroup and so on.
+        // Consequently, those first SubgroupSz^2 invocations will store their results on SubgroupSz scratch slots 
+        // with halfSubgroupSz padding and the next level will follow the same + the previous as an `offset`.
+        const uint offset = (gl_LocalInvocationIndex >> glsl::gl_SubgroupSizeLog2()) & ~subgroupMask; // For subgroupSz = 64, first 4095 invocations get offset 0, 4096-8191 offset 64, then 128 etc.
+        const uint memBegin = offset;
+        uint nextLevelStoreIndex = memBegin + glsl::gl_SubgroupID();
+        scanStoreIndex = gl_LocalInvocationIndex;
+        
+        bool participate = gl_LocalInvocationIndex <= lastInvocationInLevel;
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if (any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation)))
+                { // only invocations that have the final value of the subgroupOp (inclusive scan) store their results
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan)
+                {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+            if(isScan)
+            {
+                scanStoreIndex += lastInvocationInLevel + 1u;
+            }
+        }
+        if(lastInvocationInLevel >= glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan) {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
```

Comment:
ok just as I outlined in this series of comments: https://github.com/Devsh-Graphics-Programming/Nabla/pull/519/files#r1327248553

you want to remove the top barrier, you don't need one between a subgroup scan or get and a set

Reply:
btw I'd make the intent of this code clearer and replace the block past the barrier with
```cpp
participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
if(participate)
{
   const uint prevLevelScan = sharedAccessor.main.get(scanLoadIndex);
   scan = subgroupOp(prevLevelScan);
   scanLoadIndex += lastInvocationInLevel+1;
   if (gl_LocalInvocationIndex==0)
      accessor.set(scanLoadIndex/*loadStoreIndexDiff is 0*/,scan);
}
```

Reply:
btw its actually not necesary to perform the set at the end, unless you're broadcasting